src/test/java/com/mojang/serialization/CodecTests.java

```diff
@@ -165,6 +170,28 @@ public void list_invalidValues() {
         );
     }
 
+    @Test
+    public void sizeLimitedList_roundTrip() {
+        assertRoundTrip(
+            Codec.STRING.listOf(2),
+            List.of("foo", "bar"),
+            List.of("foo", "bar")
+        );
+    }
+
+    @Test
+    public void sizeLimitedList_tooLong() {
+        final Codec<List<String>> codec = Codec.STRING.listOf(2);
+        assertFromJavaFails(codec, List.of("foo", "bar", "baz"));
+        assertToJavaFails(codec, List.of("foo", "bar", "baz"));
+
+        // Input is clipped in partial result
+        assertEquals(
+            List.of("foo", "bar"),
+            fromJavaOrPartial(codec, List.of("foo", "bar", "baz", 4, false))
```

Comment:
Suggestion: split this test to cover both list with correct and incorrect contents?